File: Dynamic Programming/Python/Strange Printer.py
# Problem link: https://leetcode.com/problems/strange-printer/description
# Time: O(n ^ 2)
# Space: O(n ^ 2)
class Solution:
    def strangePrinter(self, s: str) -> int:
        cache = [[0] * len(s) for _ in range(len(s))]
        return self.DP(cache, 0, len(s) - 1, s)
    
    def DP(self, cache, start, end, s):
        if start == end:
            return 1
        elif cache[start][end] != 0:
            return cache[start][end]
        
        moves = inf

        for idx in range(start, end):
            moves = min(moves, self.DP(cache, start, idx, s) + self.DP(cache, idx + 1, end, s))

        cache[start][end] = moves if s[start] != s[end] else moves - 1
        return cache[start][end]

File: Dynamic Programming/Java/Strange Printer.java
// Problem link: https://leetcode.com/problems/strange-printer/description
// Time: O(n ^ 3)
// Space: O(n ^ 2)
class Solution {
    public int strangePrinter(String s) {
        int size = s.length();
        char[] str = s.toCharArray();
        int[][] cache = new int[size][size];

        return this.DP(cache, 0, size - 1, str);
    }

    private int DP(int[][] cache, int start, int end, char[] str) {
        if(start == end) {
            return 1;
        }
        else if(cache[start][end] != 0) {
            return cache[start][end];
        }
        
        int moves = Integer.MAX_VALUE;

        for(int idx = start; idx < end; idx++) {
            moves = Math.min(moves, this.DP(cache, start, idx, str) + this.DP(cache, idx + 1, end, str));
            
        }
        
        cache[start][end] = str[start] == str[end] ? moves - 1 : moves;
        return cache[start][end];
    }
}
